File: CSES/Graph-Algorithms/Labyrinth.py
n,m = map(int,input().split())
grid = [input() for i in range(n)]
 
dx = [-1,1,0,0]
dy = [0,0,-1,1]
INF = int(1e18)
 
par = [['?' for j in range(m)] for i in range(n)]
dist = [[INF for j in range(m)] for i in range(n)]
 
def is_valid(x,y):
    global n,m
    return x >= 0 and x < n and y >= 0 and y < m and grid[cx][cy] != '#'
 
pq = []
for i in range(n):
    for j in range(m):
        if grid[i][j] == 'A':
            pq.append([0, (i,j)])
            dist[i][j] = 0
            break
    if pq:
        break
 
while pq:
    cur = pq[0]
    pq.pop(0)
    if dist[cur[1][0]][cur[1][1]] != cur[0]:
        continue
    for i in range(4):
        cx = cur[1][0]+dx[i]
        cy = cur[1][1]+dy[i]
        if not is_valid(cx,cy):
            continue
        if dist[cx][cy] > cur[0]+1:
            dist[cx][cy] = cur[0]+1
            pq.append([dist[cx][cy], (cx,cy)])
            if i == 0:
                par[cx][cy] = 'U'
            elif i == 1:
                par[cx][cy] = 'D'
            elif i == 2:
                par[cx][cy] = 'L'
            else:
                par[cx][cy] = 'R'
 
 
for i in range(n):
    for j in range(m):
        if grid[i][j] == 'B':
            if dist[i][j] == INF:
                print('NO')
                exit(0)
            else:
                ans = ''
                cx,cy = i,j
                while is_valid(cx,cy) and grid[cx][cy] != 'A':
                    ans += par[cx][cy]
                    if par[cx][cy] == 'U':
                        cx += 1
                    elif par[cx][cy] == 'D':
                        cx -= 1
                    elif par[cx][cy] == 'L':
                        cy += 1
                    else:
                        cy -= 1
                print('YES')
                print(dist[i][j])
                print(ans[::-1])
                exit(0)
 

File: CSES/Graph-Algorithms/Labyrinth.cpp
#include <bits/stdc++.h>
using namespace std;
#ifdef LOCAL
#include "debugger.h"
#else
#define dbg(...) 5
#endif

#define ll long long

const int INF = 1e9;
int dx[] = {-1,1,0,0};
int dy[] = {0,0,-1,1};

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    int n, m;
    cin >> n >> m;
    vector<string> grid(n);
    for(auto& e:grid){
        cin >> e;
    }

    vector<vector<char>> par(n, vector<char>(m, '?'));
    vector<vector<int>> dist(n, vector<int>(m, INF));
    queue<pair<int, pair<int,int>>> bfs;

    for(int i = 0; i < n && bfs.empty(); i++){
        for(int j = 0; j < m; j++){
            if(grid[i][j] == 'A'){
                dist[i][j] = 0;
                bfs.push({0, {i,j}});
                break;
            }
        }
    }

    auto is_valid = [&](int x,int y){
        return x >= 0 && x < n && y >= 0 && y < m && grid[x][y] != '#';
    };

    while(!bfs.empty()){
        auto cur = bfs.front();
        bfs.pop();
        int x = cur.second.first, y = cur.second.second, d = cur.first;
        if(dist[x][y] != d){
            continue;
        }
        for(int i = 0; i < 4; i++){
            int cx = x+dx[i];
            int cy = y+dy[i];
            if(is_valid(cx,cy) && dist[cx][cy] > d+1){
                dist[cx][cy] = d+1;
                if(i == 0){
                    par[cx][cy] = 'U';
                }
                else if(i == 1){
                    par[cx][cy] = 'D';
                }
                else if(i == 2){
                    par[cx][cy] = 'L';
                }
                else{
                    par[cx][cy] = 'R';
                }
                bfs.push({dist[cx][cy], {cx,cy}});
            }
        }
    }

    for(int i = 0; i < n; i++){
        for(int j = 0; j < m; j++){
            if(grid[i][j] == 'B'){
                if(dist[i][j] == INF){
                    cout << "NO\n";
                    return 0;
                }
                cout << "YES\n" << dist[i][j] << "\n";
                string ans = "";
                int cx = i, cy = j;
                while(grid[cx][cy] != 'A'){
                    ans += par[cx][cy];
                    if(par[cx][cy] == 'U'){
                        cx++;
                    }
                    else if(par[cx][cy] == 'D'){
                        cx--;
                    }
                    else if(par[cx][cy] == 'L'){
                        cy++;
                    }
                    else{
                        cy--;
                    }
                }
                reverse(ans.begin(),ans.end());
                cout << ans << "\n";
                return 0;
            }
        }
    }
    return 0;
}
